src/analysis/schwartz.lean

```diff
@@ -0,0 +1,400 @@
+⟨λ f, ⟨-f, f.smooth.neg,
+  begin
+    intros k n,
+    rcases f.decay k n with ⟨C, hC, hf⟩,
+    use [C, hC],
+    intro x,
+    exact (decay_neg_aux k n f x).le.trans (hf x),
+  end⟩ ⟩
+
+end neg
+
+section add
+
+instance : has_add 𝓢(E, F) :=
+⟨λ f g, ⟨f + g, f.smooth.add g.smooth,
+  begin
+    intros k n,
+    rcases f.decay k n with ⟨Cf, hCf, hf⟩,
+    rcases g.decay k n with ⟨Cg, hCg, hg⟩,
+    refine ⟨Cf + Cg, by positivity, λ x, _⟩,
+    exact (decay_add_le_aux k n f g x).trans (add_le_add (hf x) (hg x)),
+  end⟩ ⟩
+
+@[simp] lemma add_apply {f g : 𝓢(E, F)} {x : E} : (f + g) x = f x + g x := rfl
+
+end add
+
+section sub
+
+instance : has_sub 𝓢(E, F) :=
+⟨λ f g, ⟨f - g, f.smooth.sub g.smooth,
+  begin
+    intros k n,
+    rcases f.decay k n with ⟨Cf, hCf, hf⟩,
+    rcases g.decay k n with ⟨Cg, hCg, hg⟩,
+    refine ⟨Cf + Cg, by positivity, λ x, _⟩,
+    refine le_trans _ (add_le_add (hf x) (hg x)),
+    rw sub_eq_add_neg,
+    rw ←decay_neg_aux k n g x,
+    convert decay_add_le_aux k n f (-g) x,
+    -- exact fails with deterministic timeout
+  end⟩ ⟩
+
+@[simp] lemma sub_apply {f g : 𝓢(E, F)} {x : E} : (f - g) x = f x - g x := rfl
+
+end sub
+
+section add_comm_group
+
+variables [normed_space ℂ F]
+
+instance : add_comm_group 𝓢(E, F) :=
+fun_like.coe_injective.add_comm_group _ rfl (λ _ _, rfl) (λ _, rfl) (λ _ _, rfl) (λ _ _, rfl)
+  (λ _ _, rfl)
+
+variables (E F)
+
+/-- Coercion as an additive homomorphism. -/
+def coe_hom : 𝓢(E, F) →+ (E → F) :=
+{ to_fun := λ f, f, map_zero' := coe_zero, map_add' := λ _ _, rfl }
+
+variables {E F}
+
+lemma coe_coe_hom : (coe_hom E F : 𝓢(E, F) → (E → F)) = coe_fn := rfl
+
+lemma coe_hom_injective : function.injective (coe_hom E F) :=
+by { rw coe_coe_hom, exact fun_like.coe_injective }
+
+end add_comm_group
+
+section module
+
+variables [normed_space ℂ F]
+variables [semiring R] [module R ℂ] [module R F] [smul_comm_class ℝ R F]
+variables [has_continuous_const_smul R F] [is_scalar_tower R ℂ F]
+
+instance : module R 𝓢(E, F) :=
+coe_hom_injective.module R (coe_hom E F) (λ _ _, rfl)
+
+end module
+
+section seminorms
+
+/-! ### Seminorms on Schwartz space-/
+
+/-- Helper definition for the seminorms of the Schwartz space. -/
+@[protected]
+def seminorm_aux (k n : ℕ) (f : 𝓢(E, F)) : ℝ :=
+Inf {c | 0 ≤ c ∧ ∀ x, ∥x∥^k * ∥iterated_fderiv ℝ n f x∥ ≤ c}
+
+lemma seminorm_aux_nonneg (k n : ℕ) (f : 𝓢(E, F)) : 0 ≤ f.seminorm_aux k n :=
+le_cInf (bounds_nonempty k n f) (λ _ ⟨hx, _⟩, hx)
+
+lemma le_seminorm_aux (k n : ℕ) (f : 𝓢(E, F)) (x : E) :
+  ∥x∥ ^ k * ∥iterated_fderiv ℝ n ⇑f x∥ ≤ f.seminorm_aux k n :=
+le_cInf (bounds_nonempty k n f) (λ y ⟨_, h⟩, h x)
+
+section
+
+open tactic tactic.positivity
+
+/-- Extension for the `positivity` tactic: seminorms are nonnegative. -/
+@[positivity]
+meta def _root_.tactic.positivity_schwartz_seminorm_aux : expr → tactic strictness
+| `(schwartz.seminorm_aux %%a %%b %%c) := nonnegative <$> mk_app ``seminorm_aux_nonneg [a, b, c]
+| _ := failed
```

Comment:
since `seminorm_aux` will only be used a few times below, and then it will be completely subsumed by the true seminorm, I am not sure it makes sense to have a positivity extension for that.

Reply:
Yes this is only for three uses of `positivity`, but is adding a positivity extension expensive? I can remove it if you don't like it, but my impression was that `positivity` was exactly for these types of things (plus writing the extension was way more interesting than writing some stupid `le_add`-term).

Reply:
Each extension will make positivity a bit slower. So for something that's not meant to be ever used again, I'd rather avoid it.